[WebDriver][BiDi] Implement the log module

lauromoura opened the pull request:
#### 66c45d7c64726e3b9d5700305bd2945c7a978e4d
<pre>
[WebDriver][BiDi] Implement the log module
<a href="https://bugs.webkit.org/show_bug.cgi?id=271956">https://bugs.webkit.org/show_bug.cgi?id=271956</a>

Reviewed by BJ Burg and Devin Rousso.

Adds the basic infrastructure for global WebDriver BiDi events,
using the `log.entryAdded` event as the first event implemented.

Console logging is supported through a new client class that gets
notified of console messages by the new `AutomationInstrumentation`,
based on the existing `InspectorInstrumentation` class.

This version supports text console messages, missing advanced
information like stacktraces, extra args and realm id.

This commit also adds back the WebAutomationSession.messages.in, so the
WebProcess can notify the UIProcess when a new console message arrives.
It was originally removed in 286516@main. It also includes follow up
fixes after the AbstractRefCounted changes from 287239@main. Note that
as we don&apos;t have a preference for WebDriver yet, these messages are
marked as ExceptionForEnabledBy.

Also, WebDriver BiDi is enabled in developer builds for WPE, so we can
actually test it on CI.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/automation/AutomationInstrumentation.cpp: Added.
(WebCore::Inspector::automationClient):
(WebCore::AutomationInstrumentation::setClient):
(WebCore::AutomationInstrumentation::clearClient):
(WebCore::logEntryLevelForMessage):
(WebCore::logEntryMethodNameForMessage):
(WebCore::AutomationInstrumentation::addMessageToConsole):
* Source/WebCore/automation/AutomationInstrumentation.h: Added.
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::PageConsoleClient::addMessage):
(WebCore::PageConsoleClient::messageWithTypeAndLevel):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::addConsoleMessage):
(WebCore::WorkerGlobalScope::addMessage):
* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session):
(WebDriver::Session::dispatchEvent):
(WebDriver::Session::doLogEntryAdded):
(WebDriver::Session::emitEvent):
(WebDriver::Session::eventIsEnabled):
(WebDriver::Session::enableGlobalEvent):
(WebDriver::Session::disableGlobalEvent):
(WebDriver::Session::toInternalEventName):
* Source/WebDriver/Session.h:
(WebDriver::Session::create):
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::dispatchMessage):
(WebDriver::SessionHost::dispatchEvent):
* Source/WebDriver/SessionHost.h:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::bidiSessionSubscribe):
(WebDriver::WebDriverService::bidiSessionUnsubscribe):
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketMessageHandler::Message::fail):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/JavaScriptCore.serialization.in:
* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::setProcessPool):
(WebKit::WebAutomationSession::logEntryAdded):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.messages.in: Copied from Source/WebKit/Shared/JavaScriptCore.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::WebAutomationSessionProxy):
(WebKit::WebAutomationSessionProxy::~WebAutomationSessionProxy):
(WebKit::WebAutomationSessionProxy::addMessageToConsole):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/288145@main">https://commits.webkit.org/288145@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d9e387e188d44c8183cb405ff3a3242a3534318

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81917 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1444 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35873 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86474 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32948 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84023 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1479 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9267 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63886 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21609 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84987 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1098 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74548 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44170 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/81370 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1000 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28725 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31368 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74898 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72333 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29334 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87905 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80971 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9159 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6529 "Found 1 new test failure: imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-no-sink.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72250 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9344 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70367 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71472 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17829 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15574 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14494 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/545 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9110 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14646 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103383 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8950 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25096 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12475 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10758 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/InspectorInstrumentation.cpp:
```diff
@@ -79,14 +80,26 @@
 namespace {
 static HashSet<InstrumentingAgents*>* s_instrumentingAgentsSet = nullptr;
+#if ENABLE(WEBDRIVER_BIDI)
+static WeakHashSet<InspectorInstrumentationConsoleMessageObserver>& consoleMessageObservers()
```
this is likely going to be used in for a lot more things than just console messages.  furthermore, i'm not sure i like the blending of Web Inspector and WebDriver things.  id much prefer if we created a new `AutomationInstrumentation` or something like that and used it in the same places that there's `InspectorInstrumentation` instead.  this way, we can adjust each independently of each other as needed.

lauromoura replied:
Right. I added a new `AutomationInstrumentation`, with just the new `addMessageToConsole`.

dcrousso commented on Source/WebCore/inspector/InspectorInstrumentation.cpp:
```diff
@@ -930,8 +958,14 @@ void InspectorInstrumentation::addMessageToConsoleImpl(InstrumentingAgents& inst
+    if (auto* consoleAgent = instrumentingAgents.webConsoleAgent()) {
+#if ENABLE(WEBDRIVER_BIDI)
+        for (const InspectorInstrumentationConsoleMessageObserver& observer : consoleMessageObservers())
+            observer(*message);
```
for example, why is this dependent on whether there's a `WebConsoleAgent`?  by this point we already have the `ConsoleMessage` so it shouldn't matter if it exists since this is for WebDriver, not Web Inspector

dcrousso commented on Source/WebCore/inspector/InspectorInstrumentation.cpp:
```diff
@@ -79,14 +80,26 @@
+#if ENABLE(WEBDRIVER_BIDI)
+static WeakHashSet<InspectorInstrumentationConsoleMessageObserver>& consoleMessageObservers()
+{
+    static NeverDestroyed<WeakHashSet<InspectorInstrumentationConsoleMessageObserver>> observers;
```
do we actually want to support more than one observer?

lauromoura replied:
Especially after moving to the `AutomationInstrumentation` scheme, a single observer will be enough. I also renamed the methods from add/remove to set/clear.

dcrousso commented on Source/WebDriver/Session.cpp:
```diff
@@ -3144,4 +3145,183 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
+    // WebDriver uses the ECMA time format, which is the number of milliseconds since the Unix epoch.
+    // https://tc39.es/ecma262/#sec-time-values-and-time-range
+    RefPtr<JSON::Value> timestampValue;
+    if (auto timestampOpt = params->getDouble("timestamp"_s))
```
i dont think we can assume that `params` exists

lauromoura replied:
Added a check.

dcrousso commented on Source/WebDriver/Session.cpp:
```diff
@@ -3144,4 +3145,183 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
+    if (auto timestampOpt = params->getDouble("timestamp"_s))
+        timestampValue = JSON::Value::create(round(*timestampOpt));
+    else
+        timestampValue = JSON::Value::create(0);
```
you could rework this to use `timestampOpt.value_or(0)` instead of having the `else`

dcrousso commented on Source/WebDriver/Session.cpp:
```diff
@@ -3144,4 +3145,183 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
+        if (arg->type() == JSON::Value::Type::String)
+            textBuilder.append(arg->asString());
+        else
+            textBuilder.append("null"_s);
```
are we not supposed to serialize boolean or number values?

lauromoura replied:
Yes, there's a lengthy spec about serialization of the arguments. But after reviewing the `Automation.json` and the `WebAutomationSession` messages, they send only the final text string for a given console message.

So, for the sake of simplicity, we can remove the serialization steps _from this PR_, adding back in the follow-up PR's regarding multiple arguments/formatter strings (https://bugs.webkit.org/show_bug.cgi?id=282976) and actual serialization (https://bugs.webkit.org/show_bug.cgi?id=282977).

Maybe the serialization/formatting can be actually be implemented in the UIProcess instead of in the WebDriver service. @burg had some ideas in this direction for other BiDi methods/events too.

dcrousso commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2689,6 +2691,40 @@ void WebDriverService::bidiSessionStatus(unsigned id, RefPtr<JSON::Object>&&, Fu
+    // https://bugs.webkit.org/show_bug.cgi?id=282436
+    // FIXME: Support by-context subscriptions.
+    // https://bugs.webkit.org/show_bug.cgi?id=282981
+    for (auto& eventName : *eventNames) {
```
we cant assume that `eventNames` exists

dcrousso commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -2689,6 +2691,40 @@ void WebDriverService::bidiSessionStatus(unsigned id, RefPtr<JSON::Object>&&, Fu
+
+    // FIXME: Support by-context unsubscriptions.
+    // https://bugs.webkit.org/show_bug.cgi?id=282981
+    for (auto& eventName : *eventNames) {
```
ditto

burg commented on Source/WebCore/workers/WorkerGlobalScope.cpp:
```diff
@@ -82,6 +82,9 @@
 
+#if ENABLE(WEBDRIVER_BIDI)
+#include "AutomationInstrumentation.h"
+#endif
```
Nit: newline after #endif

lauromoura replied:
Thanks for the review. I just addressed Devin's comments, before actually merging the PR.

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.h:
```diff
@@ -0,0 +1,68 @@
+namespace WebCore {
+
+
+using AutomationInstrumentationConsoleMessageObserver = WTF::Observer<void(const Inspector::ConsoleMessage&)>;
```
instead of having this be a separate object, could we create an abstract class that `WebAutomationSessionProxy` inherits from?
```cpp
class AutomationInstrumentationClient {
public:
    virtual ~AutomationInstrumentationClient() = default;

    virtual addMessageToConsole(const std::unique_ptr<Inspector::ConsoleMessage>&) { }
};
```

i just find it kinda unnecessary to have something specific to console messages when it's extremely likely that there's gonna be other things `WebAutomationSessionProxy` will want to know about (i.e. it'd not be idea if the same `WebAutomationSessionProxy` is provided as an observer for a dozen different events)

lauromoura replied:
Good catch. I added this client class in place of the observer.

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.h:
```diff
@@ -0,0 +1,68 @@
+
+inline void AutomationInstrumentation::addMessageToConsole(const std::unique_ptr<Inspector::ConsoleMessage>& message)
+{
+    addMessageToConsoleImpl(message);
```
im not really sure what benefit this extra level of indirection provides

for `InspectorInstrumentation` we will quickly return if there are no frontends present.  perhaps we should do the same if there are no automation sessions?

lauromoura replied:
I replaced this indirection with a direct implementation, given it's a "shallow" method.

dcrousso commented on Source/WebDriver/Session.cpp:
```diff
@@ -3144,4 +3146,139 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
+        }
+
+        if (capitalizeNext) {
+            builder.append(static_cast<char>(u_toupper(eventName[i])));
```
all event names should be ASCII right?  if so, i think we could use `toASCIIUpper` instead

lauromoura replied:
The spec does not explicitly enforce this (at least I could not find), but the currently defined (and supported) events are ascii, indeed.

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.h:
```diff
@@ -0,0 +1,72 @@
+
+inline void AutomationInstrumentationClient::addMessageToConsole(const Inspector::ConsoleMessage&)
+{
+}
```
NIT: i'd just inline this above

lauromoura replied:
Do you mean moving this right after the `AutomationInstrumentionClient` declaration, before the `AutomationInstrumentation` class?

dcrousso replied:
i mean having it be right when you first declare it
```cpp
static void addMessageToConsole(const std::unique_ptr<Inspector::ConsoleMessage>&) { }
```

lauromoura replied:
Oh, right. I accidentally had left something inside the header declaration, which the compiler complained about. I'll update it.

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.cpp:
```diff
@@ -0,0 +1,68 @@
+
+void AutomationInstrumentation::addMessageToConsole(const std::unique_ptr<ConsoleMessage>& message)
+{
+    if (s_client)
```
can we at least mark this as `if (UNLIKELY(s_client))` just in case that helps make things faster?

im a bit wary of not having something like `FAST_RETURN_IF_NO_FRONTENDS` (or in this case `FAST_RETURN_IF_NO_CLIENT`) since this will be called on _every_ `console` message.  have you done any performance measurements to see if this has an impact?

lauromoura replied:
Not yet, I'll check it.

lauromoura replied:
So, I did some "unscientific" benchmarks, measuring the time to call console.log 100k times between measurements, with 2k such iterations on each run.

It showed similar results with/without `UNLIKELY`, both in the mean/media/stddev, and in the 25/75 quartiles of the samples.

In any case, I'll add it as it's the expected behavior.

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.cpp:
```diff
@@ -0,0 +1,68 @@
+using namespace Inspector;
+
+namespace {
+static WeakPtr<AutomationInstrumentationClient> s_client;
```
should this be wrapped in a function instead?
```cpp
static WeakPtr<AutomationInstrumentationClient>& client() {
    static NeverDestroyed<WeakPtr<AutomationInstrumentationClient>> client;
    return client.get();
}
```

lauromoura replied:
I kept it separate to make it easier to clear it from the `clearClient()` function, used when we destroy the `WebAutomationSessionProxy`.

dcrousso replied:
im not sure i follow.  you can do this instead and it should work the same as what you have now
```cpp
client() = nullptr;
```

dcrousso commented on Source/WebCore/workers/WorkerGlobalScope.cpp:
```diff
@@ -466,6 +470,9 @@ void WorkerGlobalScope::addConsoleMessage(std::unique_ptr<Inspector::ConsoleMess
         PageConsoleClient::logMessageToSystemConsole(*message);
 
+#if ENABLE(WEBDRIVER_BIDI)
+    AutomationInstrumentation::addMessageToConsole(message);
```
this might not happen on the main thread.  is that a problem?  or is `WebAutomationSessionProxy` threadsafe?

lauromoura replied:
IIUC, `WebAutomationSessionProxy` is not thread-safe by itself (e.g. it inherits from `RefCounted` instead of `ThreadSafeRefCounted`, etc.).

But its `addMessageToConsole` method just forwards its parameters to the respective `WebAutomationSession` on the UIProcess through `WebProcess::singleton()->parentProcessConnection()->send(...)`, without accessing other instance data. In the header, `send()` is documented as thread-safe. Is this enough?

Maybe in the future, when instrumenting something else, `WebAutomationSessionProxy` will actually need to be made thread-safe?

dcrousso replied:
but is `WebProcess::singleton` threadsafe?

perhaps just in case you could have a `ensureOnMainThread` somewhere?

lauromoura replied:
That's an option.

Given `Inspector::ConsoleMessage` is noncopyable and `WebAutomationSession::addMessageToConsole` (the client function) already unpacks the information from it before passing ahead to the `UIProcess`, I'll update the patch to make `AutomationInstrumentation::addMessageToConsole` extract the information, on the thread where we get it, passing to the main thread (and further down the pipeline) a simplified object, with just the streamlined information we need. Does this sound ok?

dcrousso commented on Source/WebCore/workers/WorkerGlobalScope.cpp:
```diff
@@ -486,6 +493,10 @@ void WorkerGlobalScope::addMessage(MessageSource source, MessageLevel level, con
         message = makeUnique<Inspector::ConsoleMessage>(source, MessageType::Log, level, messageText, sourceURL, lineNumber, columnNumber, state, requestIdentifier);
+
+#if ENABLE(WEBDRIVER_BIDI)
+    AutomationInstrumentation::addMessageToConsole(message);
```
ditto

lauromoura replied:
ditto answer above

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.cpp:
```diff
@@ -0,0 +1,111 @@
+    auto method = logEntryMethodNameForMessage(*message);
+    WTF::ensureOnMainThread([level, text, type, timestamp, method] {
+        if (LIKELY(automationClient()))
+            automationClient()->addMessageToConsole(level, text, type, timestamp, method);
```
i think it actually makes more sense to send the typed values instead of strings since it's possible some other client might want to do things with them in code
```cpp
WTF::ensureOnMainThread([source = message.source(), type = message.type(), level = message.level(), text = message.message(), timestamp = message.timestamp()] {
    if (LIKELY(automationClient()))
        automationClient()->addMessageToConsole(source, type, level, text, timestamp);
});
```
the client should be the one responsible for converting the typed values into strings, and probably only right before it puts them into the JSON since it's much cheaper/easier/etc. to send typed values over IPC than strings

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2569,6 +2576,17 @@ Ref<WebAutomationSession::Debuggable> WebAutomationSession::protectedDebuggable(
+
+    // FIXME Get browsing context handle and source info
+    // https://bugs.webkit.org/show_bug.cgi?id=282981
+    m_domainNotifier->logEntryAdded(level, sourceString, text, timestamp.secondsSinceEpoch().milliseconds(), type, method.value_or(emptyString()));
```
so i think it'd be here that does the conversion from `MessageSource`/`MessageType`/`MessageLevel` to `String`

we could also have enum types for each inside the `Automation` domain, but perhaps that's not necessary for now

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.cpp:
```diff
@@ -0,0 +1,79 @@
+void AutomationInstrumentation::setClient(const AutomationInstrumentationClient &client)
+{
+    ASSERT(!automationClient());
+    automationClient() = &client;
```
NIT: do we need to `= &client` or can it just be `= client`?  im pretty sure `WeakPtr<T>` can be created from `T&`

lauromoura replied:
`= client` was enough.

dcrousso commented on Source/WebCore/automation/AutomationInstrumentation.cpp:
```diff
@@ -0,0 +1,79 @@
+
+    WTF::ensureOnMainThread([source = message->source(), type = message->type(), level = message->level(), messageText = message->message(), timestamp = message->timestamp()] {
+        if (LIKELY(automationClient()))
+            automationClient()->addMessageToConsole(source, level, messageText, type, timestamp);
```
i think with the new pointer safety rules we actually need to do something like this
```cpp
if (RefPtr client = automationClient().get())
    client->addMessageToConsole(source, level, messageText, type, timestamp);
```

dcrousso reviewed (approved):
thanks for iterating with me :)

can we add any sort of tests for this to make sure it's at least minimally correct?  like one for a page that `console.log()` and then another that does the same from a `Worker`?

lauromoura committed:
[WebDriver][BiDi] Implement the log module
https://bugs.webkit.org/show_bug.cgi?id=271956

Reviewed by BJ Burg and Devin Rousso.

Adds the basic infrastructure for global WebDriver BiDi events,
using the `log.entryAdded` event as the first event implemented.

Console logging is supported through a new client class that gets
notified of console messages by the new `AutomationInstrumentation`,
based on the existing `InspectorInstrumentation` class.

This version supports text console messages, missing advanced
information like stacktraces, extra args and realm id.

This commit also adds back the WebAutomationSession.messages.in, so the
WebProcess can notify the UIProcess when a new console message arrives.
It was originally removed in 286516@main. It also includes follow up
fixes after the AbstractRefCounted changes from 287239@main. Note that
as we don't have a preference for WebDriver yet, these messages are
marked as ExceptionForEnabledBy.

Also, WebDriver BiDi is enabled in developer builds for WPE, so we can
actually test it on CI.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/automation/AutomationInstrumentation.cpp: Added.
(WebCore::Inspector::automationClient):
(WebCore::AutomationInstrumentation::setClient):
(WebCore::AutomationInstrumentation::clearClient):
(WebCore::logEntryLevelForMessage):
(WebCore::logEntryMethodNameForMessage):
(WebCore::AutomationInstrumentation::addMessageToConsole):
* Source/WebCore/automation/AutomationInstrumentation.h: Added.
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::PageConsoleClient::addMessage):
(WebCore::PageConsoleClient::messageWithTypeAndLevel):
* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::addConsoleMessage):
(WebCore::WorkerGlobalScope::addMessage):
* Source/WebDriver/Session.cpp:
(WebDriver::Session::Session):
(WebDriver::Session::dispatchEvent):
(WebDriver::Session::doLogEntryAdded):
(WebDriver::Session::emitEvent):
(WebDriver::Session::eventIsEnabled):
(WebDriver::Session::enableGlobalEvent):
(WebDriver::Session::disableGlobalEvent):
(WebDriver::Session::toInternalEventName):
* Source/WebDriver/Session.h:
(WebDriver::Session::create):
* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::dispatchMessage):
(WebDriver::SessionHost::dispatchEvent):
* Source/WebDriver/SessionHost.h:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::bidiSessionSubscribe):
(WebDriver::WebDriverService::bidiSessionUnsubscribe):
* Source/WebDriver/WebDriverService.h:
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketMessageHandler::Message::fail):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/JavaScriptCore.serialization.in:
* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::setProcessPool):
(WebKit::WebAutomationSession::logEntryAdded):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.messages.in: Copied from Source/WebKit/Shared/JavaScriptCore.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::WebAutomationSessionProxy):
(WebKit::WebAutomationSessionProxy::~WebAutomationSessionProxy):
(WebKit::WebAutomationSessionProxy::addMessageToConsole):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/288145@main
